<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>genotype frequencies â€¢ hypogen</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="genotype frequencies">
<meta property="og:description" content="hypogen">
<meta property="og:image" content="https://k-hench.github.io/hypogen/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hypogen</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/hypogen.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/hypogen_genotype_frequncies.html">genotype frequencies</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/k-hench/hypogen/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="hypogen_genotype_frequncies_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>genotype frequencies</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/k-hench/hypogen/blob/master/vignettes/hypogen_genotype_frequncies.Rmd"><code>vignettes/hypogen_genotype_frequncies.Rmd</code></a></small>
      <div class="hidden name"><code>hypogen_genotype_frequncies.Rmd</code></div>

    </div>

    
    
<style type="text/css">
img {
  border: 0;
}
</style>
<div id="preparation" class="section level2">
<h2 class="hasAnchor">
<a href="#preparation" class="anchor"></a>Preparation:</h2>
<p>To be able to plot genotytpe frequencies, we need a genotype table where genotypes are encoded as 0,1, &amp; 2 (homozygous reference allele, heterozygous and homozygous alternative allele).</p>
<p>This can be easily generated from a <strong>vcf file</strong> using <strong>VCFtools</strong>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="ex">vcftools</span> \</span>
<span id="cb1-2"><a href="#cb1-2"></a>  --gzvcf input.vcf.gz \</span>
<span id="cb1-3"><a href="#cb1-3"></a>  --out  genotype_table \</span>
<span id="cb1-4"><a href="#cb1-4"></a>  --012</span></code></pre></div>
<p>Unfortunately the <strong>vcftools</strong> output has the wrong orientation (samples as rows, sites as columns). So before we reed the genotypes into <strong>R</strong> they are transformed using <strong>awk</strong>:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="fu">awk</span> <span class="st">'</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="st">{</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="st">    for (i=1; i&lt;=NF; i++)  {</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="st">        a[NR,i] = $i</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="st">    }</span></span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="st">}</span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="st">NF&gt;p { p = NF }</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="st">END {</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="st">    for(j=1; j&lt;=p; j++) {</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="st">        str=a[1,j]</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="st">        for(i=2; i&lt;=NR; i++){</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="st">            str=str" "a[i,j];</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="st">        }</span></span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="st">        print str</span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="st">    }</span></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="st">}'</span> genotype_table.012 <span class="kw">|</span> <span class="fu">gzip</span> <span class="op">&gt;</span> genotype_table.012.trans.txt.gz</span></code></pre></div>
</div>
<div id="r-setup" class="section level2">
<h2 class="hasAnchor">
<a href="#r-setup" class="anchor"></a>R setup</h2>
<p>To efficiently plot the genotypes frequencies we make use of the <strong>ggtern</strong> package as well as of the <strong>tidyverse</strong> (and of the <strong>hypogen</strong> package):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.ggtern.com">ggtern</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="loading-the-data-into-r" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-the-data-into-r" class="anchor"></a>Loading the data into R</h2>
<p>The package <strong>hypogen</strong> provides the function <code><a href="../reference/hypo_import_genotype_freq.html">hypo_import_genotype_freq()</a></code> which can read the transformed genotypee table and directly computes genotype frequencies:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">file_genotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"genotype_table.012.trans.txt.gz"</span>, package <span class="op">=</span> <span class="st">"hypogen"</span><span class="op">)</span>

<span class="va">genotype_freqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hypo_import_genotype_freq.html">hypo_import_genotype_freq</a></span><span class="op">(</span><span class="va">file_genotypes</span><span class="op">)</span></code></pre></div>
<p>This table can be directly plotted using <strong>gggtern</strong>. As a reference we also make use of the <strong>hypogen</strong> function <code><a href="../reference/hypo_hwe.html">hypo_hwe()</a></code> to indicate Hardy-Weinberg genotype frequencies:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/ggtern/man/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtern/man/coord_tern.html">coord_tern</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="va">genotype_freqs</span>,
             <span class="fu"><a href="https://rdrr.io/pkg/ggtern/man/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">AA</span>, y <span class="op">=</span> <span class="va">Aa</span>, z <span class="op">=</span> <span class="va">aa</span> <span class="op">)</span>,
             col <span class="op">=</span> <span class="va">hypo_clr_LGs</span><span class="op">[</span><span class="fl">18</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_path</span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="../reference/hypo_hwe.html">hypo_hwe</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>,
            <span class="fu"><a href="https://rdrr.io/pkg/ggtern/man/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">AA</span>, y <span class="op">=</span> <span class="va">Aa</span>, z <span class="op">=</span> <span class="va">aa</span> <span class="op">)</span>,
            color <span class="op">=</span><span class="st">'black'</span>, alpha <span class="op">=</span> <span class="fl">.6</span><span class="op">)</span></code></pre></div>
<p><img src="hypogen_genotype_frequncies_files/figure-html/plot-1.png" width="100%"></p>
<hr>
<p><img src="hypogen_genotype_frequncies_files/figure-html/unnamed-chunk-2-1.png" width="150pt" height="150pt" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://k-hench.github.io/kh/">Kosmas Hench</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
